controllers/remote/cluster_cache.go

```diff
@@ -54,319 +48,206 @@ const (
-// kind, and eventHandler, then this is a no-op.
-func (m *ClusterCacheTracker) Watch(ctx context.Context, input WatchInput) error {
-	gvk, err := apiutil.GVKForObject(input.Kind, m.scheme)
+// newAccessor creates a new accessor.
+func (t *ClusterCacheTracker) newAccessor(ctx context.Context, cluster client.ObjectKey) (*accessor, error) {
+	// Get a rest config for the remote cluster
+	config, err := RESTConfig(ctx, t.client, cluster)
 	if err != nil {
-		return err
-	}
-
-	wi := watchInfo{
-		watcher:               input.Watcher,
-		gvk:                   gvk,
-		eventHandlerSignature: eventHandlerSignature(input.EventHandler),
+		return nil, errors.Wrap(err, "error fetching REST client config for remote cluster")
 	}
+	config.Timeout = defaultClientTimeout
 
-	// First, check if the watch already exists
-	var exists bool
-	m.watchesLock.RLock()
-	exists = m.watchExists(input.Cluster, wi)
-	m.watchesLock.RUnlock()
-
-	if exists {
-		m.log.V(4).Info("Watch already exists", "namespace", input.Cluster.Namespace, "cluster", input.Cluster.Name, "kind", fmt.Sprintf("%T", input.Kind))
-		return nil
+	// Create a mapper for it
+	mapper, err := apiutil.NewDynamicRESTMapper(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating dynamic rest mapper for remote cluster")
 	}
 
-	// Doesn't exist - grab the write lock
-	m.watchesLock.Lock()
-	defer m.watchesLock.Unlock()
-
-	// Need to check if another goroutine created the watch while this one was waiting for the lock
-	if m.watchExists(input.Cluster, wi) {
-		m.log.V(4).Info("Watch already exists", "namespace", input.Cluster.Namespace, "cluster", input.Cluster.Name, "kind", fmt.Sprintf("%T", input.Kind))
-		return nil
+	// Create the client for the remote cluster
+	c, err := client.New(config, client.Options{Scheme: t.scheme, Mapper: mapper})
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating client for remote cluster")
 	}
 
-	// Need to create the watch
-	watchesForCluster, found := m.watches[input.Cluster]
-	if !found {
-		watchesForCluster = make(map[watchInfo]struct{})
-		m.watches[input.Cluster] = watchesForCluster
+	// Create the cache for the remote cluster
+	cacheOptions := cache.Options{
+		Scheme: t.scheme,
+		Mapper: mapper,
 	}
-
-	cache, err := m.getOrCreateClusterCache(ctx, input.Cluster)
+	remoteCache, err := cache.New(config, cacheOptions)
 	if err != nil {
-		return err
+		return nil, errors.Wrap(err, "error creating cache for remote cluster")
 	}
 
-	if err := input.Watcher.Watch(source.NewKindWithCache(input.Kind, cache), input.EventHandler, input.Predicates...); err != nil {
-		return errors.Wrap(err, "error creating watch")
+	// We need to be able to stop the cache's shared informers, so wrap this in a stoppableCache.
+	cache := &stoppableCache{
+		Cache: remoteCache,
+		stop:  make(chan struct{}),
 	}
 
-	watchesForCluster[wi] = struct{}{}
-
-	return nil
-}
+	// Start the cache!!!
+	go remoteCache.Start(cache.stop)
```

Comment:
I find the mixed usage or `remoteCache` and `cache` kind of confusing. Is it possible to always use `cache` her and below in this func?

Reply:
Let me see what I can do.